packages/labs/src/select.tsx

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+
+import {
+    AbstractComponent,
+    Button,
+    Classes,
+    HTMLInputProps,
+    IInputGroupProps,
+    InputGroup,
+    IPopoverProps,
+    Keys,
+    Popover,
+    Position,
+    Utils,
+} from "@blueprintjs/core";
+import { IInputListRendererProps, IListItemsProps, InputList } from "./inputList";
+
+export interface ISelectProps<T> extends IListItemsProps<T> {
+    /**
+     * Whether the dropdown list can be filtered.
+     * Disabling this option will remove the `InputGroup` and ignore `inputProps`.
+     * @default true
+     */
+    filterable?: boolean;
+
+    /**
+     * Custom renderer for an item in the dropdown list. Receives a boolean indicating whether
+     * this item is active (selected by keyboard arrows) and an `onClick` event handler that
+     * should be attached to the returned element.
+     */
+    itemRenderer: (item: T, isActive: boolean, onClick: React.MouseEventHandler<HTMLElement>) => JSX.Element;
+
+    /** React child to render when filtering items returns zero results. */
+    noResults?: string | JSX.Element;
+
+    /**
+     * Props to spread to `InputGroup`. All props are supported except `ref` (use `inputRef` instead).
+     * If you want to control the filter input, you can pass `value` and `onChange` here
+     * to override `Select`'s own behavior.
+     */
+    inputProps?: IInputGroupProps & HTMLInputProps;
+
+    /** Props to spread to `Popover`. Note that `content` cannot be changed. */
+    popoverProps?: Partial<IPopoverProps> & object;
+
+    /**
+     * Whether the filtering state (query string and active item) should be
+     * reset to initial values when an item is selected.
+     * @default false
+     */
+    resetOnSelect?: boolean;
+}
+
+export interface ISelectState<T> {
+    activeItem?: T;
+    isOpen?: boolean;
```

Comment:
`isOpen` and `query` are never undefined. Should they be optional? or is it the blueprint's style to make all state fields optional?